### PR TITLE
[WIP] Preliminary work splitting const qualification into separate passes

### PR DESCRIPTION
**Pull request by alexreg**
r? @eddyb

CC @oli-obk

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -101,7 +101,9 @@ struct Qualifier<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     rpo: ReversePostorder<'a, 'tcx>,
     tcx: TyCtxt<'a, 'gcx, 'tcx>,
     param_env: ty::ParamEnv<'tcx>,
-    local_qualif: IndexVec<Local, Option<Qualif>>,
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: The commit message isn't up to date anymore. The temp + return merge is already in master

File: src/test/compile-fail/const-block-non-item-statement.rs [hunk 2]
```diff
@@ -10,10 +10,17 @@
 
 enum Foo {
     Bar = { let x = 1; 3 }
+<<<<<<< HEAD
     //~^ ERROR let bindings in constants are unstable
     //~| ERROR statements in constants are unstable
     //~| ERROR let bindings in constants are unstable
     //~| ERROR statements in constants are unstable
+||||||| merged common ancestors
```
Comment (oli-obk) [inline src/test/compile-fail/const-block-non-item-statement.rs]: rebase leftovers

File: src/librustc/mir/interpret/error.rs [hunk 3]
```diff
@@ -175,6 +176,7 @@ pub enum EvalErrorKind<'tcx, O> {
     NoMirFor(String),
     UnterminatedCString(Pointer),
     DanglingPointerDeref,
+    ExternStaticRead(DefId),
```
Comment (oli-obk) [inline src/librustc/mir/interpret/error.rs]: This is introduced here and removed two commits later. Merge the commits into each other?

File: src/test/compile-fail/const-block-non-item-statement.rs (hunk 2 above)
Comment (alexreg) [reply src/test/compile-fail/const-block-non-item-statement.rs]: You'd think Git wouldn't let me do this... grr....

File: src/librustc_mir/transform/qualify_consts.rs [hunk 4]
```diff
@@ -199,46 +201,45 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
...
+        let store = |this: &mut Self, index| {
+            this.local_mut_interior.set_member(&index, this.qualif.mut_interior);
+            this.local_needs_drop.set_member(&index, this.qualif.needs_drop);
+            this.local_not_const.set_member(&index, this.qualif.not_const);
         };
 
         // Only handle promotable temps in non-const functions.
         if self.mode == Mode::Fn {
             if let Place::Local(index) = *dest {
                 if self.mir.local_kind(index) == LocalKind::Temp
```
Comment (oli-obk) [inline src/librustc_mir/transform/qualify_consts.rs]: I believe the failure [here](https://github.com/rust-lang/rust/blob/master/src/test/run-pass/issue-49955.rs#L19) is due to this not also entering this arm in case of `LocalKind::ReturnPointer`. Although I don't see how that worked before!?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 5]
```diff
@@ -398,14 +471,377 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
...
+
+/// Accumulates an Rvalue or a Call's effects into `self.mut_interior`.
+impl<'a, 'tcx> Visitor<'tcx> for MutInteriorChecker<'a, 'tcx, 'tcx> {
+    fn visit_local(&mut self,
+                   &local: &Local,
+                   _: PlaceContext<'tcx>,
+                   _: Location) {
+        let kind = self.mir.local_kind(local);
+        match kind {
+            LocalKind::ReturnPointer => {}
```
Comment (alexreg) [inline src/librustc_mir/transform/qualify_consts.rs]: @oli-obk @eddyb Should the arm for this actually be something like `self.mut_interior |= self.local_mut_interior.contains(&RETURN_PLACE);`?

**Comment by alexreg** [comment]
@eddyb Lots of weird errors like ```can't find crate for `` ``` and such. Any idea?

**Comment by pietroalbini** [comment]
Ping from triage @eddyb! The author needs some help.

File: src/librustdoc/lib.rs [hunk 6]
```diff
@@ -310,19 +310,19 @@ pub fn opts() -> Vec<RustcOptGroup> {
...
-        unstable("deny", |o| {
+        stable("deny", |o| {
             o.optmulti("D", "deny", "Set lint denied", "OPT")
         }),
-        unstable("forbid", |o| {
+        stable("forbid", |o| {
             o.optmulti("F", "forbid", "Set lint forbidden", "OPT")
         }),
-        unstable("cap-lints", |o| {
+        stable("cap-lints", |o| {
```
Comment (eddyb) [inline src/librustdoc/lib.rs]: Why this change?

File: src/librustc_data_structures/indexed_set.rs [hunk 7]
```diff
@@ -210,6 +210,16 @@ impl<T: Idx> IdxSet<T> {
         self.bits.set_bit(elem.index())
     }
 
+    /// Adds `elem` to the set `self` if `member` is `true`, otherwise removes `elem` from the set;
+    /// returns true iff this changed `self`.
```
Comment (eddyb) [inline src/librustc_data_structures/indexed_set.rs]: Do you need this? Alternatively, you have `BitVector` (yes, we ended up with 2 APIs for the same purpose).

File: src/librustc_mir/transform/qualify_consts.rs [hunk 8]
```diff
@@ -35,54 +36,28 @@ use syntax::ast::LitKind;
...
-        }
-        if !ty.needs_drop(tcx, param_env) {
-            *self = *self - Qualif::NEEDS_DROP;
-        }
+    if needs_drop {
+        bits |= NEEDS_DROP_BITS;
     }
+    if not_const {
+        bits |= NOT_CONST_BITS;
+    }
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: You can simplify this by using an enum `Bits` and then:
```rust
(mut_interior as u8) << (Bits::MutInterior as u32) |
(needs_drop as u8) << (Bits::NeedsDrop as u32) |
(not_const as u8) << (Bits::NotConst as u32)
```

File: src/librustc_mir/transform/qualify_consts.rs [hunk 9]
```diff
@@ -106,61 +81,180 @@ impl fmt::Display for Mode {
...
     mir: &'a Mir<'tcx>,
     rpo: ReversePostorder<'a, 'tcx>,
     tcx: TyCtxt<'a, 'gcx, 'tcx>,
     param_env: ty::ParamEnv<'tcx>,
-    local_qualif: IndexVec<Local, Option<Qualif>>,
-    qualif: Qualif,
+    local_not_const: IdxSetBuf<Local>,
+    not_const: bool,
+    mut_interior_checker: MutInteriorChecker<'a, 'gcx, 'tcx>,
+    drop_checker: DropChecker<'a, 'gcx, 'tcx>,
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Oh you're doing them in lockstep. Makes sense, as a first approximation.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 10]
```diff
@@ -106,61 +81,180 @@ impl fmt::Display for Mode {
...
-            local_qualif,
-            qualif: Qualif::empty(),
+            local_not_const,
+            not_const: false,
+            mut_interior_checker,
+            drop_checker,
             const_fn_arg_vars: BitVector::new(mir.local_decls.len()),
             temp_promotion_state: temps,
-            promotion_candidates: vec![]
+            promotion_candidates: Vec::new()
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Why this change? AFAIK we prefer `vec![]`.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 11]
```diff
@@ -199,46 +293,27 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
...
-    /// be combined with the value before the call to nest.
+    /// be combined with the value before the call to `nest`.
     fn nest<F: FnOnce(&mut Self)>(&mut self, f: F) {
-        let original = self.qualif;
-        self.qualif = Qualif::empty();
+        let original = self.not_const;
+        self.not_const = false;
         f(self);
-        self.add(original);
+        self.not_const |= original;
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Pretty sure this is broken because it doesn't *actually* nest the other two.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 12]
```diff
@@ -199,46 +293,27 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
...
     }
 
     /// Assign the current qualification to the given destination.
     fn assign(&mut self, dest: &Place<'tcx>, location: Location) {
         trace!("assign: {:?}", dest);
-        let qualif = self.qualif;
-        let span = self.span;
-        let store = |slot: &mut Option<Qualif>| {
-            if slot.is_some() {
-                span_bug!(span, "multiple assignments to {:?}", dest);
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: You removed this sanity check.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 13]
```diff
@@ -249,22 +324,20 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
                                    self.mir.local_kind(index) == LocalKind::Arg) &&
                                    self.tcx.sess.features_untracked().const_let => {
                 debug!("store to var {:?}", index);
-                self.local_qualif[index] = Some(self.qualif);
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Hmm you're losing the distinction between uninitialized and no flags set.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 14]
```diff
@@ -106,61 +81,180 @@ impl fmt::Display for Mode {
...
-           -> Qualifier<'a, 'tcx, 'tcx> {
+           -> ConstChecker<'a, 'tcx, 'tcx> {
         let mut rpo = traversal::reverse_postorder(mir);
         let temps = promote_consts::collect_temps(mir, &mut rpo);
         rpo.reset();
 
         let param_env = tcx.param_env(def_id);
 
-        let mut local_qualif = IndexVec::from_elem(None, &mir.local_decls);
+        let mut local_not_const = IdxSetBuf::new_filled(mir.local_decls.len());
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: Ah, clever trick!

File: src/librustc_mir/transform/qualify_consts.rs [hunk 15]
```diff
@@ -319,7 +392,9 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
                         // Check for unused values. This usually means
                         // there are extra statements in the AST.
                         for temp in mir.temps_iter() {
-                            if self.local_qualif[temp].is_none() {
+                            if !self.mut_interior_checker.local_mut_interior.contains(&temp) &&
+                               !self.drop_checker.local_needs_drop.contains(&temp) &&
+                               !self.local_not_const.contains(&temp) {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: This replacement is wrong. What `None` meant was "uninitialized", whereas this is checking for "no flags set" aka "promotable".

File: src/librustc_mir/transform/qualify_consts.rs [hunk 16]
```diff
@@ -398,55 +471,410 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
...
+
+/// Accumulates an Rvalue or a Call's effects into `self.mut_interior`.
+impl<'a, 'tcx> Visitor<'tcx> for MutInteriorChecker<'a, 'tcx, 'tcx> {
+    fn visit_local(&mut self,
+                   &local: &Local,
+                   _: PlaceContext<'tcx>,
+                   _: Location) {
+        let kind = self.mir.local_kind(local);
+        match kind {
+            LocalKind::Var if !self.tcx.sess.features_untracked().const_let => {}
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: You don't need to check features here or handle cases the "not const" checker errors from, specially.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 11 above)
Comment (eddyb) [reply src/librustc_mir/transform/qualify_consts.rs]: Well, it could work, but *only* if the other two aren't accessed from this one.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 17]
```diff
@@ -922,7 +1327,9 @@ This does not pose a problem by itself because they can't be accessed directly."
                         // this doesn't come from a macro that has #[allow_internal_unstable]
                         !self.span.allows_unstable()
                     {
-                        self.qualif = Qualif::NOT_CONST;
+                        self.mut_interior_checker.mut_interior = false;
+                        self.drop_checker.needs_drop = false;
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: You probably don't need to touch the other flags here.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 18]
```diff
@@ -342,7 +417,9 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
                         }
 
                         // Make sure there are no extra unassigned variables.
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: I'm not sure this loop even works anymore?

File: src/librustc_mir/transform/qualify_consts.rs [hunk 19]
```diff
@@ -106,61 +81,180 @@ impl fmt::Display for Mode {
...
+            drop_checker,
             const_fn_arg_vars: BitVector::new(mir.local_decls.len()),
             temp_promotion_state: temps,
-            promotion_candidates: vec![]
+            promotion_candidates: Vec::new()
         }
     }
 
+    /// Returns whether all qualification flags are unset.
+    fn is_qualif_empty(&self) -> bool {
```
Comment (eddyb) [inline src/librustc_mir/transform/qualify_consts.rs]: This is a hazard because it checks the other two flags without capturing the "instantaneous state".

**Comment by TimNN** [comment]
Ping from triage, @alexreg. It looks like some changes have been requested to your PR.

**Comment by alexreg** [comment]
@TimNN Yeah, I was away on holiday, but just getting to them now. Hopefully will have them resolved in the next day or two. :-)

File: src/librustdoc/lib.rs (hunk 6 above)
Comment (alexreg) [reply src/librustdoc/lib.rs]: Err, I didn't even realise I made this! 😮

Comment (alexreg) [reply src/librustdoc/lib.rs]: Local Git tells me I haven't even modified this file.

File: src/librustc_data_structures/indexed_set.rs (hunk 7 above)
Comment (alexreg) [reply src/librustc_data_structures/indexed_set.rs]: Yeah, it's used. I don't see how it's immediately similar to `BitVector`...

File: src/librustc_mir/transform/qualify_consts.rs (hunk 8 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Good point!

File: src/librustc_mir/transform/qualify_consts.rs (hunk 10 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Do you? I always prefer the function myself... if it's convention or in guidelines though, I can change it.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 11 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Yep, you're right... the other two are accessed.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 13 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Yep. What do you recommend? You're the one who suggested moving to `IdxSetBuf` instead of `BitVector` if I remember...

File: src/librustc_mir/transform/qualify_consts.rs (hunk 18 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Nor am I...

File: src/librustc_mir/transform/qualify_consts.rs (hunk 17 above)
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Are you sure? I did a literal translation of course, because I wasn't sure...

File: src/librustc_mir/transform/qualify_consts.rs [hunk 20]
```diff
@@ -1035,40 +1435,23 @@ This does not pose a problem by itself because they can't be accessed directly."
...
     fn visit_source_info(&mut self, source_info: &SourceInfo) {
+        self.mut_interior_checker.visit_source_info(source_info);
+        self.drop_checker.visit_source_info(source_info);
+
         self.span = source_info.span;
     }
 
     fn visit_statement(&mut self, bb: BasicBlock, statement: &Statement<'tcx>, location: Location) {
+        self.mut_interior_checker.visit_statement(bb, statement, location);
+        self.drop_checker.visit_statement(bb, statement, location);
```
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_mir/transform/qualify_consts.rs]: Right. So I need to do this at the granularity of MIR "nodes" ("elements"?) – i.e. per `visit` method?

**Comment by mark-i-m** [comment]
@Aaronepower 

It looks like @eddyb is needed: https://github.com/rust-lang/rust/issues/49146#issuecomment-414168991

**Comment by Aaronepower** [comment]
@mark-i-m Thank you, I've updated the status of the PR to reflect this.

**Comment by alexreg** [comment]
@Aaronepower Yep. I've been pestering @eddyb about this for a while. I think he'll finally get around to it soon. :-)

**Comment by pietroalbini** [comment]
Ping from triage @eddyb! This PR needs your review.

**Comment by TimNN** [comment]
Ping from triage @eddyb / @rust-lang/compiler: This PR requires your review.

**Comment by TimNN** [comment]
Ping from triage @eddyb / @rust-lang/compiler: This PR requires your review.

**Comment by alexreg** [comment]
@TimNN Appreciate your efforts to try to get @eddyb to review this... I've spoken with him on Discord though, and he's too busy with 2018 Edition stuff at this moment. That said, a week today (or tomorrow) he should be able to tackle this! :-)

**Comment by RalfJung** [comment]
Conflict ahead: https://github.com/rust-lang/rust/pull/54424

File: src/librustdoc/lib.rs (hunk 6 above)
Comment (RalfJung) [reply src/librustdoc/lib.rs]: Probably it happened during a rebase then? Still, you should remove this change.

**Comment by TimNN** [comment]
Ping from triage @eddyb / @alexreg: What is the status of this PR?

**Comment by alexreg** [comment]
@eddyb Still waiting on @eddyb... he's been AWOL for a little while now.

**Comment by TimNN** [comment]
Ping from triage @eddyb / @rust-lang/compiler: What are your plans for this PR?

**Comment by estebank** [comment]
_Side note_: please do not leave merge commits in the PR.

**Comment by alexreg** [comment]
@estebank That's a rebase gone wrong I think. :-P

**Comment by TimNN** [comment]
Ping from triage @eddyb: This PR requires your review!

**Comment by Centril** [comment]
Ping from triage @eddyb / @rust-lang/compiler  -- we are eagerly awaiting your review.

**Comment by alexreg** [comment]
Closing this PR in favour of @eddyb's new one https://github.com/rust-lang/rust/pull/58403.
